core/math/random_pcg.cpp

```diff
@@ -55,7 +55,7 @@ int64_t RandomPCG::rand_weighted(const Vector<float> &p_weights) {
 	float remaining_distance = randf() * weights_sum;
 	for (int64_t i = 0; i < weights_size; ++i) {
 		remaining_distance -= weights[i];
-		if (remaining_distance < 0) {
+		if (remaining_distance <= 0) {
```

Comment:
This can lead to returning an index of a zero weight, before the change it could not happen. E.g. consider:
- All weights being `0.0`. Now it would return `0` instead of `-1`.
- The first weight being `0.0`. When `randf()` will return `0.0` then the return value will be `0` as well.

Aka it fixes the case when `randf()` returns `1.0` but it breaks some other cases.

---

Also note that passing negative weights is not prevented anyhow. I think it's fine to not handle it at all / assume it's invalid input, but maybe it should be mentioned in the docs that passing negative weights will result in undefined behavior? :thinking:

However zero weights absolutely make sense to me and I see no reason to make them be not handled correctly.

So I think we could change it back to:
```suggestion
		if (remaining_distance < 0) {
```
and then after the loop we could return the index of the last non-zero weight, and `-1` if there's no such weight.

Reply:
I didn't think it through this far, you're right. Returning the index of the last non-zero weight at the end of the loop makes sense and seems to be the better solution